### ISSUE TITLE
undue State Error raised in Cisco IOS XE 17.9.x show tacacs output when using cisco_ios_show_tacacs.textfsm

### ISSUE THREAD

**Issue by mcn29** [networktocode/ntc-templates#1890]
##### ISSUE TYPE
 - Template Issue with error and raw data

##### TEMPLATE USING
ntc_templates/templates/cisco_ios_show_tacacs.textfsm
```
Value TACACS_SERVER_NAME (.+?)
Value TACACS_SERVER (.+?)
Value SERVER_PORT (\d+)
Value SOCKET_OPENS (\d+)
Value SOCKET_CLOSES (\d+)
Value SOCKET_ABORTS (\d+)
Value SOCKET_ERRORS (\d+)
Value SOCKET_TIMEOUTS (\d+)
Value FAILED_CONNECTIONS (\d+)
Value PACKETS_SENT (\d+)
Value PACKET_RECEIVED (\d+)


Start
  ^Tacacs -> Record
  ^\s+Server\s+address:\s+${TACACS_SERVER}\s*$$
  ^\s+Server\s+name:\s+${TACACS_SERVER_NAME}\s*$$
  ^\s+Server\s+port:\s+${SERVER_PORT}\s*$$
  ^\s+Socket\s+opens:\s+ ${SOCKET_OPENS}\s*$$
  ^\s+Socket\s+closes:\s+${SOCKET_CLOSES}\s*$$
  ^\s+Socket\s+aborts:\s+${SOCKET_ABORTS}\s*$$
  ^\s+Socket\s+errors:\s+${SOCKET_ERRORS}\s*$$
  ^\s+Socket\s+Timeouts:\s+${SOCKET_TIMEOUTS}\s*$$
  ^\s+Failed\s+Connect\s+Attempts:\s+${FAILED_CONNECTIONS}\s*$$
  ^\s+Total\s+Packets\s+Sent:\s+${PACKETS_SENT}\s*$$
  ^\s+Total\s+Packets\s+Recv:\s+${PACKET_RECEIVED}\s*$$
  ^\s*$$
  # Capture time-stamp if vty line has command time-stamping turned on
  ^Load\s+for\s+
  ^Time\s+source\s+is
  ^. -> Error
```

##### SAMPLE COMMAND OUTPUT

```

Tacacs+ Server -  public  :
               Server name: TACACS_1
            Server address: 10.10.10.1
               Server port: 49
              Socket opens:      21752
             Socket closes:      21752
             Socket aborts:          0
             Socket errors:          0
           Socket Timeouts:          0
   Failed Connect Attempts:          0
        Total Packets Sent:      23450
        Total Packets Recv:      23439
             Server Status: Alive
Continous Authc fail count:          2
Continous Authz fail count:          0


Tacacs+ Server -  public  :
               Server name: TACACS_2
            Server address: 10.10.10.2
               Server port: 49
              Socket opens:        169
             Socket closes:        169
             Socket aborts:          0
             Socket errors:          0
           Socket Timeouts:          0
   Failed Connect Attempts:          0
        Total Packets Sent:        174
        Total Packets Recv:        174
             Server Status: Alive
Continous Authc fail count:          0
Continous Authz fail count:          0

```

##### SUMMARY
In a recent IOS XE software update (I use 17.9.5), Cisco introduced those 3 new lines in the output of the show tacacs command, which are not recognised by the template :

```
             Server Status: Alive
Continous Authc fail count:          0
Continous Authz fail count:          0
```
below

```
        Total Packets Recv:          0
```

##### STEPS TO REPRODUCE
Add the following config lines on any IOS XE version 17.9
```
aaa new-model
tacacs server TACACS_1
 address ipv4 10.10.10.1
 key 7 06572B6907773A44502430
tacacs server TACACS_2
 address ipv4 10.10.10.2
 key 7 06572B6907773A44502430
```
My python version 3.9 script :
```
from netmiko import ConnectHandler
connection = ConnectHandler(
                                        device_type="cisco_ios",
                                        host='10.10.10.100,
                                        username='my_user',
                                        password='my_password',
                                        )

connection.send_command('show tacacs', use_textfsm=True)
```

##### EXPECTED RESULTS

I was expecting a valid output

##### ACTUAL RESULTS
```
Traceback (most recent call last):
  File "/usr/lib64/python3.11/runpy.py", line 198, in _run_module_as_main
    return _run_code(code, main_globals, None,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/runpy.py", line 88, in _run_code
    exec(code, run_globals)
  File "/home/my_uid/.vscode-server/extensions/ms-python.debugpy-2024.1.10362351/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/home/my_uid/.vscode-server/extensions/ms-python.debugpy-2024.1.10362351/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/home/my_uid/.vscode-server/extensions/ms-python.debugpy-2024.1.10362351/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 284, in run_file
    runpy.run_path(target, run_name="__main__")
  File "/home/my_uid/.vscode-server/extensions/ms-python.debugpy-2024.1.10362351/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 321, in run_path
    return _run_module_code(code, init_globals, run_name,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/.vscode-server/extensions/ms-python.debugpy-2024.1.10362351/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 135, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/home/my_uid/.vscode-server/extensions/ms-python.debugpy-2024.1.10362351/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/home/my_uid/scripts/my_script.py", line 7301, in <module>
    (lan_acc_swi_audited, acc_po, acc_dom,dev_acc_ko) = audit_usr_stack(site,rah_user,
                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/scripts/my_script.py", line 5196, in audit_usr_stack
    check_tacacs(hostname,net_conn_acc,run_beg,'acc',dsn_int,res_lst,acc_remed_report)
  File "/home/my_uid/scripts/my_script.py", line 2205, in check_tacacs
    run_tacacs=connection.send_command('show tacacs', use_textfsm=True)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/netmiko/base_connection.py", line 111, in wrapper_decorator
    return_val = func(self, *args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/netmiko/utilities.py", line 596, in wrapper_decorator
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/netmiko/base_connection.py", line 1834, in send_command
    return_val = structured_data_converter(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/netmiko/utilities.py", line 560, in structured_data_converter
    structured_output_tfsm = get_structured_data_textfsm(
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/netmiko/utilities.py", line 380, in get_structured_data_textfsm
    output = _textfsm_parse(textfsm_obj, raw_output, attrs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/netmiko/utilities.py", line 344, in _textfsm_parse
    tfsm_parse(raw_output, attrs)
  File "/home/my_uid/venv/lib64/python3.11/site-packages/textfsm/clitable.py", line 282, in ParseCmd
    self.table = self._ParseCmdItem(self.raw, template_file=template_files[0])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/textfsm/clitable.py", line 315, in _ParseCmdItem
    for record in fsm.ParseText(cmd_input):
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/textfsm/parser.py", line 895, in ParseText
    self._CheckLine(line)
  File "/home/my_uid/venv/lib64/python3.11/site-packages/textfsm/parser.py", line 944, in _CheckLine
    if self._Operations(rule, line):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/my_uid/venv/lib64/python3.11/site-packages/textfsm/parser.py", line 1024, in _Operations
    raise TextFSMError('State Error raised. Rule Line: %s. Input Line: %s'
textfsm.parser.TextFSMError: State Error raised. Rule Line: 31. Input Line:              Server Status: Alive
```

**Comment by mcn29** [comment]
This is how I got round the issue : I added the code that takes those 3 new lines into account :
```
             Server Status: Alive
Continous Authc fail count:          0
Continous Authz fail count:          0
```

But I'm not sure that fix will suit most as some may not want to have 3 new keys:values (STATUS , AUTHC_FAIL , AUTHZ_FAIL ) in the output. If it does, the names of the new keys will need to be discussed.

```
Value TACACS_SERVER_NAME (.+?)
Value TACACS_SERVER (.+?)
Value SERVER_PORT (\d+)
Value SOCKET_OPENS (\d+)
Value SOCKET_CLOSES (\d+)
Value SOCKET_ABORTS (\d+)
Value SOCKET_ERRORS (\d+)
Value SOCKET_TIMEOUTS (\d+)
Value FAILED_CONNECTIONS (\d+)
Value PACKETS_SENT (\d+)
Value PACKET_RECEIVED (\d+)
Value STATUS (.+?)
Value AUTHC_FAIL (\d+)
Value AUTHZ_FAIL (\d+)

Start
  ^Tacacs -> Record
  ^\s+Server\s+address:\s+${TACACS_SERVER}\s*$$
  ^\s+Server\s+name:\s+${TACACS_SERVER_NAME}\s*$$
  ^\s+Server\s+port:\s+${SERVER_PORT}\s*$$
  ^\s+Socket\s+opens:\s+ ${SOCKET_OPENS}\s*$$
  ^\s+Socket\s+closes:\s+${SOCKET_CLOSES}\s*$$
  ^\s+Socket\s+aborts:\s+${SOCKET_ABORTS}\s*$$
  ^\s+Socket\s+errors:\s+${SOCKET_ERRORS}\s*$$
  ^\s+Socket\s+Timeouts:\s+${SOCKET_TIMEOUTS}\s*$$
  ^\s+Failed\s+Connect\s+Attempts:\s+${FAILED_CONNECTIONS}\s*$$
  ^\s+Total\s+Packets\s+Sent:\s+${PACKETS_SENT}\s*$$
  ^\s+Total\s+Packets\s+Recv:\s+${PACKET_RECEIVED}\s*$$
  ^\s+Server\s+Status:\s+${STATUS}\s*$$
  ^Continous\s+Authc\s+fail\s+count:\s+${AUTHC_FAIL}\s*$$
  ^Continous\s+Authz\s+fail\s+count:\s+${AUTHZ_FAIL}\s*$$
  ^\s*$$
  # Capture time-stamp if vty line has command time-stamping turned on
  ^Load\s+for\s+
  ^Time\s+source\s+is
  ^. -> Error
```

**Comment by mjbear** [comment]
@mcn29
I was curious what the status would look like if I added a bogus server. Oddly enough, the bogus T+ server continues to have a status of "Alive". oooof :shaking_face:

If the Server Status and Continuous Auth lines aren't of much interest then add a line to match 'em without bothering to capture data.

I'm game to team up and work with you. :grinning:

**Comment by mcn29** [comment]
> @mcn29 I was curious what the status would look like if I added a bogus server. Oddly enough, the bogus T+ server continues to have a status of "Alive". oooof 🫨
> 
> If the Server Status and Continuous Auth lines aren't of much interest then add a line to match 'em without bothering to capture data.
> 
> I'm game to team up and work with you. 😀

@mjbear 
Hi Michael,
Yes, I had noticed the status was always 'Alive'. I'm not sure about the reason. Either we need to wait for the "tacacs-server deadtime" available in the nexus range or we need to type some aaa commands to test the status.

One more reason not to capture the data.

Anyway, if we choose to ignore the 3 new lines, this is what I've come up with below.

From my understanding, when those lines are not found, which will occur on older IOS versions, then the processing will carry on.

```
Value TACACS_SERVER_NAME (.+?)
Value TACACS_SERVER (.+?)
Value SERVER_PORT (\d+)
Value SOCKET_OPENS (\d+)
Value SOCKET_CLOSES (\d+)
Value SOCKET_ABORTS (\d+)
Value SOCKET_ERRORS (\d+)
Value SOCKET_TIMEOUTS (\d+)
Value FAILED_CONNECTIONS (\d+)
Value PACKETS_SENT (\d+)
Value PACKET_RECEIVED (\d+)


Start
  ^Tacacs -> Record
  ^\s+Server\s+address:\s+${TACACS_SERVER}\s*$$
  ^\s+Server\s+name:\s+${TACACS_SERVER_NAME}\s*$$
  ^\s+Server\s+port:\s+${SERVER_PORT}\s*$$
  ^\s+Socket\s+opens:\s+ ${SOCKET_OPENS}\s*$$
  ^\s+Socket\s+closes:\s+${SOCKET_CLOSES}\s*$$
  ^\s+Socket\s+aborts:\s+${SOCKET_ABORTS}\s*$$
  ^\s+Socket\s+errors:\s+${SOCKET_ERRORS}\s*$$
  ^\s+Socket\s+Timeouts:\s+${SOCKET_TIMEOUTS}\s*$$
  ^\s+Failed\s+Connect\s+Attempts:\s+${FAILED_CONNECTIONS}\s*$$
  ^\s+Total\s+Packets\s+Sent:\s+${PACKETS_SENT}\s*$$
  ^\s+Total\s+Packets\s+Recv:\s+${PACKET_RECEIVED}\s*$$
  ^\s+Server\s+Status:.+?\s*$$
  ^Continous\s+Authc\s+fail\s+count:\s+\d+\s*$$
  ^Continous\s+Authz\s+fail\s+count:\s+\d+\s*$$
  ^\s*$$
  # Capture time-stamp if vty line has command time-stamping turned on
  ^Load\s+for\s+
  ^Time\s+source\s+is
  ^. -> Error
```
Marc

**Comment by mjbear** [comment]
> @mjbear Hi Michael, Yes, I had noticed the status was always 'Alive'. I'm not sure about the reason. Either we need to wait for the "tacacs-server deadtime" available in the nexus range or we need to type some aaa commands to test the status.
> 

Hello Marc @mcn29,

When I was doing my testing, I waited what I felt was long enough for the T+ attempts to drop into a failed/dead state. I also set the timeout much lower on the aaa group to see if that would help.

I had used a decommissioned ASR1k running 17.9.4 to test.

> One more reason not to capture the data.
> 
> Anyway, if we choose to ignore the 3 new lines, this is what I've come up with below.
> 
> From my understanding, when those lines are not found, which will occur on older IOS versions, then the processing will carry on.

You're correct, those lines for 17.9.x will not pose an issue for older IOS.

Spin up a PR and I'd be glad to review!

**Comment by mcn29** [comment]
Hello Michael @mjbear,
PR #1891
Marc

**Comment by mjbear** [comment]
@jmcgill298 
This is resolved by that #1891 we were having so much fun with. :face_with_peeking_eye:
Thank you!
